apps/aecore/src/aec_conductor.erl

```diff
@@ -947,7 +947,8 @@ setup_loop(State = #state{ consensus = Cons }, RestartMining, IsLeader, Origin)
     State1 = State#state{ consensus = Cons#consensus{ leader = IsLeader } },
     State2 =
         case Origin of
-            block_created when IsLeader ->
+            Origin when IsLeader, Origin =:= block_created
+                           orelse Origin =:= block_received ->
                 aec_block_generator:start_generation(),
```

Comment:
I am checking two things:
* What is effect of calling `aec_block_generator:start_generation()` and what is expected state. (I need to read code.)
* Where code branches on whether this block received from HTTP internal API is the mined version of the block candidate. Specifically:
  * Whether signing pubkey (miner) in received key block is same for which priv key is known.
  * Whether rest of the received block fits with internal state e.g. state (for building microblock on correct state trees), timestamp (for correct timing in emission of microblocks). It feels simply the received key block shall be equal to the known block candidate, or (in a more sophisticated manner) in a pool of known block candidates (so to have the priv key).

I have the impression that this `block_received` tag shall be split in two subtags for the following code paths depending on whether this received key block is meant to start a generation of microblocks.

(I may be missing a check that is already in place, checking.)

Reply:
* `aec_block_generator:start_generation()` 👍 
* There is no such check and I'm not really sure we need one. Should end user be able to post valid through the api? How different it is from receiving one from sync? Anyway I thing the decision possible changes should be mede in another PR. 
For key blocks `IsLeader` flag is set up based based comparing signing keys and block miner. We're not discarding block with different miner

Re splitting: maybe instead of using `post_block/1` in the api implementation we should go through mining code path: [`handle_mining_reply({{ok, {Nonce, Evd}}, HeaderBin}, #state{} = State) ->`](https://github.com/aeternity/epoch/blob/PT-161715282-post-key_block/apps/aecore/src/aec_conductor.erl#L688)

Reply:
> There is no such check and I'm not really sure we need one

I'm going to confirm it with original PT story for this endpoint

cc @jur0

Reply:
`aec_keys` keeps (alongside peer keypair):
* (A) The keypair currently used for signing keyblocks (`#worker_state.sign_priv` - see `aec_keys:sign_micro_block/1`).
* (B) The keypair meant to be used in key block candidate, for miner field of key block candidate (`#worker_state.candidate_pub` - see `aec_keys:candidate_pubkey/`).

[**If** key block candidate mined **and** it becomes top](https://github.com/aeternity/epoch/blob/8d69cfd2/apps/aecore/src/aec_conductor.erl#L536), candidate keypair (B) is promoted to sign keypair (A) (hence following microblocks signed with promoted keypair). Beware that conditional is combined effect of:
* [`aec_keys:promote_candidate/1` matching specified pubkey against known candidate pubkey (notice two function clauses), and returning value (as opposed to terminating)](https://github.com/aeternity/epoch/blob/8d69cfd2/apps/aecore/src/aec_keys.erl#L386-L392).
* [Call not checking return value](https://github.com/aeternity/epoch/blob/8d69cfd2/apps/aecore/src/aec_conductor.erl#L536).

So there it looks to me that there is a check - I need to understand effect. (I am considering whether only test enhancement needed.)

Reply:
> **If** key block candidate mined (EDIT: regardless of the origin, i.e. regardless of whether it mined by node or posted to node from HTTP internal API) **and** it becomes top

In same case, [conductor triggers computation of new key block candidate](https://github.com/aeternity/epoch/blob/8d69cfd2/apps/aecore/src/aec_conductor.erl#L538) after [having updated `top_block_hash` in state](https://github.com/aeternity/epoch/blob/8d69cfd2/apps/aecore/src/aec_conductor.erl#L518) that later [passes it to `aec_block_key_candidate:create`
https://github.com/aeternity/epoch/blob/8d69cfd2/apps/aecore/src/aec_conductor.erl#L804) (that fetches block hash from DB).

So key block candidate generation seems fine. I need to check the microblock generation.